components/p3a/p3a_message.cc

```diff
@@ -5,65 +5,84 @@
-  const std::string yoi = base::NumberToString(exploded.year).substr(2, 4);
+void MessageMetainfo::Init(PrefService* local_state,
+                           std::string channel,
+                           std::string week_of_install) {
+  platform = brave_stats::GetPlatformIdentifier();
+  this->channel = channel;
+  version = version_info::GetBraveVersionWithoutChromiumMajorVersion();
+
+  if (!week_of_install.empty()) {
+    date_of_install = brave_stats::GetYMDAsDate(week_of_install);
+  } else {
+    date_of_install = base::Time::Now();
+  }
+  woi = brave_stats::GetIsoWeekNumber(date_of_install);
+
+  country_code = base::ToUpperASCII(base::CountryCodeForCurrentTimezone());
+  refcode = local_state->GetString(kReferralPromoCode);
+  MaybeStripRefcodeAndCountry();
+
+  Update();
+
+  VLOG(2) << "Message meta: " << platform << " " << channel << " " << version
+          << " " << woi << " " << wos << " " << country_code << " " << refcode;
+}
+
+void MessageMetainfo::Update() {
+  date_of_survey = base::Time::Now();
+  wos = brave_stats::GetIsoWeekNumber(date_of_survey);
+}
 
-  const std::string metastring =
-      "," + meta.country_code + "," + meta.platform + "," + meta.version + "," +
-      meta.channel + "," + yoi + base::NumberToString(meta.woi) + "," + yos +
-      base::NumberToString(meta.wos) + "," + meta.refcode + ",";
+void MessageMetainfo::MaybeStripRefcodeAndCountry() {
+  const std::string& country = country_code;
+  constexpr char kRefcodeNone[] = "none";
+  constexpr char kCountryOther[] = "other";
 
-  const std::string metric_value_str = base::NumberToString(metric_value);
+  static base::flat_set<std::string> const kLinuxCountries(
+      {"US", "FR", "DE", "GB", "IN", "BR", "PL", "NL", "ES", "CA", "IT", "AU",
+       "MX", "CH", "RU", "ZA", "SE", "BE", "JP"});
 
-  // TODO(iefremov): replace with 'if'?
-  CHECK_LE(metastring.size() + metric_value_str.size(), kDataLength - 1);
+  static base::flat_set<std::string> const kNotableCountries(
+      {"US", "FR", "PH", "GB", "IN", "DE", "BR", "CA", "IT", "ES", "NL", "MX",
+       "AU", "RU", "JP", "PL", "ID", "KR", "AR"});
 
-  memcpy(ptr, metastring.data(), metastring.size());
-  ptr += metastring.size();
-  memcpy(ptr, metric_value_str.data(), metric_value_str.size());
+  // Always strip the refcode.
+  // We no longer need to partition P3A data with that key.
+  refcode = kRefcodeNone;
 
-  // Init the message.
-  p3a_message->set_metric_id(metric_hash);
-  p3a_message->set_p3a_info(data, kDataLength);
+  if (platform == "linux-bc") {
+    // If we have more than 3/0.05 = 60 users in a country for
+    // a week of install, we can send country.
+    if (kLinuxCountries.count(country) == 0) {
+      country_code = kCountryOther;
+    }
+  } else {
+    // Now the minimum platform is MacOS at ~3%, so cut off for a group under
+    // here becomes 3/(0.05*0.03) = 2000.
+    if (kNotableCountries.count(country) == 0) {
+      country_code = kCountryOther;
+    }
+  }
```

Comment:
This looks like it's the same as what it used to be.

cc privacy team (couldn't find the GH group to @ them) to check to make sure this aligns with our current privacy policies still